### ISSUE TITLE
Optimizing tensorflow model size

### ISSUE THREAD

**Issue by sebastientourbier** [Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit#52]
It is currently impossible to publish pymialsrtk to PyPi due to a 100MB limit size, mainly due to size of trained Tensorflow models (2 x ~90MB).
 
Here is a list of methods to can help in optimizing the Tensorflow models for serving prediction:

* **Freezing:** Convert the variables stored in a checkpoint file of the SavedModel into constants stored directly in the model graph. This reduces the overall size of the model.

* **Pruning:** Strip unused nodes in the prediction path and the outputs of the graph, merging duplicate nodes, as well as cleaning other node ops like summary, identity, etc.

* **Constant folding:** Look for any sub-graphs within the model that always evaluate to constant expressions, and replace them with those constants. Folding batch norms: Fold the multiplications introduced in batch normalization into the weight multiplications of the previous layer.

* **Quantization:** Convert weights from floating point to lower precision, such as 16 or 8 bits.

**References**

* https://medium.com/google-cloud/optimizing-tensorflow-models-for-serving-959080e9ddbf
* https://stackoverflow.com/questions/51957336/how-to-properly-reduce-the-size-of-a-tensorflow-savedmodel

**Comment by sebastientourbier** [comment]
@hamzake Here is the issue related to our discussion and the different options we can use to optimize the size of the trained models along with the references.

**Comment by sebastientourbier** [comment]
```python
g = create_graph()

im = np.zeros((1, width, height, n_channels))
pred3d = []
with tf.Session(graph=g) as sess_test_loc:
    # Restore the model
    tf_saver = tf.train.Saver()
    tf_saver.restore(sess_test_loc, modelCkptLoc)
    saved_path = tf_saver.save(sess_test_loc, ''.join([modelCkptLoc,'v2']))
    print('model saved in {}'.format(saved_path))
```

**Comment by hamzake** [comment]
```python
saver = tf.train.Saver(max_to_keep=1000, keep_checkpoint_every_n_hours=1)
..
if step2 % 1000 == 0:
                ## Save the variables to disk.
                saver.save(sess, model_path_save, global_step=step2)
```

**Comment by hamzake** [comment]
Dice computation:
import sklearn.metrics as skm
print("3D Dice: %.4f" % skm.f1_score(np.asarray(lbl3d).ravel().astype('int'), np.asarray(pred3d).ravel()))

**Comment by sebastientourbier** [comment]
it seems as we rely on tflearn way it is impossible or very challenging to optimize models as we can with tensorflow models for serving prediction.  So I do not think it is worst to dedicate time here. I think we should proceed with a first publication of pymialsrtk on PyPI where we do not include the checkpoints (meaning we comment the package_data line in the `setup.py` before making the `v2.0.1` release)

**Comment by sebastientourbier** [comment]
Could be reopened when we migrate to keras.
